pyiceberg/io/pyarrow.py

```diff
@@ -1308,6 +1309,192 @@ def _read_all_delete_files(fs: FileSystem, tasks: Iterable[FileScanTask]) -> Dic
     return deletes_per_file
 
 
+def _fs_from_file_path(file_path: str, io: FileIO) -> FileSystem:
+    scheme, netloc, _ = _parse_location(file_path)
+    if isinstance(io, PyArrowFileIO):
+        return io.fs_by_scheme(scheme, netloc)
+    else:
+        try:
+            from pyiceberg.io.fsspec import FsspecFileIO
+
+            if isinstance(io, FsspecFileIO):
+                from pyarrow.fs import PyFileSystem
+
+                return PyFileSystem(FSSpecHandler(io.get_fs(scheme)))
+            else:
+                raise ValueError(f"Expected PyArrowFileIO or FsspecFileIO, got: {io}")
+        except ModuleNotFoundError as e:
+            # When FsSpec is not installed
+            raise ValueError(f"Expected PyArrowFileIO or FsspecFileIO, got: {io}") from e
+
+
+class PyArrowProjector:
+    _table_metadata: TableMetadata
+    _io: FileIO
+    _fs: FileSystem
+    _projected_schema: Schema
+    _bound_row_filter: BooleanExpression
+    _case_sensitive: bool
+    _limit: Optional[int]
+    """Projects an Iceberg Table to a PyArrow construct.
```

Comment:
is "project" the right word to use here? i think "project" sounds like schema projection. this is doing schema projection, row filter, and limits

Reply:
Good point... What are good alternatives?

Reply:
This is how it is currently used,
```
        return PyArrowProjector(
            self.table_metadata, self.io, self.projection(), self.row_filter, self.case_sensitive, self.limit
        ).project_table(self.plan_files())
```
```
        batches = PyArrowProjector(
            self.table_metadata, self.io, self.projection(), self.row_filter, self.case_sensitive, self.limit
        ).project_batches(self.plan_files())
```

What about `PyArrowProjector.to_arrow_table`/`PyArrowProjector.to_record_batches`? It's what duckdb uses https://duckdb.org/docs/guides/python/export_arrow.html

but also no strong opinions here :)

Reply:
I feel that "Py" is unnecessary, and "project" in "project_batches" seems redundant.

I like @kevinjqliu's suggestion, and I want to offer another option. Perhaps the schema details could be moved to another class called `ArrowSchemaProjection`, and here, where we currently have the scanning projection, it could be renamed to `ArrowScanProjection`. The methods could then be named `for_batches` or `for_table`, like this: `ArrowScanProjection(...).for_batches` / `ArrowScanProjection(...).for_table`.

What do you think?

Reply:
I like it, make sense to me since the function `to_arrow` is in the `DataScan(TableScan)` class. `ArrowScan.to_table`/`ArrowScan.to_record_batches`

Naming is hard...

Reply:
Thank you for the suggestions @ndrluis and @kevinjqliu ! You are both such word wizards. I will use ArrowScan as the class name with the suggested method names. Thank you!